src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs

```diff
@@ -241,15 +242,30 @@ private Glyph GetGlyph(SymbolKind kind, Accessibility? declaredAccessibility)
                             var name = rule.NamingStyle.CreateName(baseName).EscapeIdentifier(context.IsInQuery);
                             if (name.Length > 1 && !result.ContainsKey(name)) // Don't add multiple items for the same name
                             {
-                                var targetToken = context.TargetToken;
-                                var uniqueName = semanticFactsService.GenerateUniqueName(
-                                    context.SemanticModel,
-                                    context.TargetToken.Parent,
-                                    containerOpt: null,
-                                    baseName: name,
-                                    filter: IsRelevantSymbolKind,
-                                    usedNames: Enumerable.Empty<string>(),
-                                    cancellationToken: cancellationToken);
+                                var location = context.TargetToken.Parent;
+
+                                // Get all symbols visible to the current location and filter to include relevant kinds.
+                                var visibleSymbols = context.SemanticModel.LookupSymbols(location.SpanStart).Where(s => IsRelevantSymbolKind(s));
+
+                                // Find additional local symbols from the current block that could cause conflict errors.
+                                // For example, we shouldn't suggest classA at $$ in the following:
+                                //     public void M()
+                                //     {
+                                //         ClassA $$
+                                //         if (true)
+                                //         {
+                                //             ClassA classA;
+                                //         }
+                                //     }
+                                // Local functions must be excluded since 
+                                //     a) nothing inside should be visible and
+                                //     b) it is valid to re-use names defined inside the local function.
+                                var container = location.AncestorsAndSelf().FirstOrDefault(
+                                    a => syntaxFactsService.IsExecutableBlock(a) || syntaxFactsService.IsMethodBody(a));
+                                var symbolsInBlock = context.SemanticModel.GetExistingSymbols(container, cancellationToken,
+                                    filter: (SyntaxNode n) => !n.IsKind(SyntaxKind.LocalFunctionStatement));
```

Comment:
this all seems like something that GenerateUniqueName should do. That's basically teh point of it.  You say "here's the name I'd like, and i'm in this location" and it contains all the smarts.

Reply:
Generate unique name does something similar, the main difference being it doesn't look at local functions.  

GenerateUniqueName is used for both VB and C#, so I can't directly move this check into there since as far as I know VB doesn't have local functions.  

I can have the logic in the declaration name completion provider (as it is), or I could potentially pass another filter to [GenerateUniqueName](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/LanguageServices/SemanticsFactsService/AbstractSemanticFactsService.cs,51) and pass it along to semanticModel.GetExistingSymbols(...)

@CyrusNajmabadi any thoughts on which is better?

Reply:
> GenerateUniqueName is used for both VB and C#, so I can't directly move this check into there since as far as I know VB doesn't have local functions.

You def can update it.  it's already in an inheritance hierarchy in case there's some VB or C# specific behavior that is needed.

Reply:
Also... this logic doesn't seem correct to me...  you're filteirng out local functions.  But it's certainly the case that a local function has a name, and you shouldn't introduce a variable with the same name as that local function.

Reply:
looking at the bug you're trying to fix, the root cause is that we're taking the declared symbols for the parameters and treating it as if that means those names are rserved in other scopes.  that's really not how local function parameters work.  First, it's acceptable for there to be a collision with them and outer variables (i.e. local function parameters were cahnged to hide outer variables).  Second, local function params are only in scope inside the local funciton, not in the containing block (like how other symbols work).  We really need to fixup GenerateUniqueName (or whatever helper it is using) to understand this.

Reply:
> 
> 
> > GenerateUniqueName is used for both VB and C#, so I can't directly move this check into there since as far as I know VB doesn't have local functions.
> 
> You def can update it. it's already in an inheritance hierarchy in case there's some VB or C# specific behavior that is needed.

Hmm it looks possible, I'll have to mess around with it a bit.  Both the the C# and VB implementation delegate to the same implementation in [AbstractSemanticFactsService](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/LanguageServices/SemanticsFactsService/AbstractSemanticFactsService.cs,8e7bbf7378df5ee8,references)

To get C# specific behavior, I'll need to make the methods AbstractSemanticFactsService virtual and then override them in the C# implementation.  I'll try that out.

Reply:
Note: once you do, please do a FAR on GenerateUniqueName to find the feautres that use it.  Then try to write at least one test per feature that uses local-functions/parameters in a sensible manner.  It will help ensure that we haven't regressed anything for those features (and perhaps will show that we've fixed a bug for them where they used to do things incorrectly).

Reply:
Thanks!